### PR TITLE
Core fm module

### PR DESCRIPTION
**Pull request by j143-zz**
For test!

### CODE REVIEW

File: scripts/staging/fm.dml [hunk 1]
```diff
@@ -0,0 +1,103 @@
+#-------------------------------------------------------------
+#
```
Comment (j143-zz) [inline scripts/staging/fm.dml:2]: Let's finalize the gradients  formulation.. please find eq(1), eq(4), lemma (3.1) in [paper](https://www.csie.ntu.edu.tw/~b97053/paper/Rendle2010FM.pdf). 
```
n -> no. of examples
d -> no. of features
k -> no. of dimensions
```

File: scripts/staging/fm.dml [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+   *  - X : n examples with d features, of shape (n, d).
+   *  - w0: the global bias, of shape (1,).
+   *  - W : the strength of each feature, of shape (d, 1).
+   *  - V : factorized interaction terms, of shape (d, k).
+   *  
+   * Outputs:
+   *  - out : target vector, of shape (n, 1)
+   */
+   
+   out = w0 + X %*% W + 0.5 * rowSums( (X %*% V)^2 - (X^2 %*% V^2) );
```
Comment (j143) [inline scripts/staging/fm.dml]:
```
w0 -> should be (n,1)
```
```
X %*% W  -> shape (n, 1)
```
```
rowSums( (X %*% V)^2 - (X^2 %*% V^2) ) -> shape (n, 1)
```

File: scripts/staging/fm.dml [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    *  - dX : the gradient of loss function w.r.t  X, of shape (n, d).
+    *  - dw0: the gradient of loss function w.r.t w0, of shape (1,).
+    *  - dW : the gradient of loss function w.r.t  W, of shape (d, 1). 
+    *  - dV : the gradient of loss function w.r.t  V, of shape (d, k).
+    */
+    
+    # gradient of y (model) w.r.t. X (input)
+    dX  = ( W + 0.5 * rowSums( (2*X %*% V^2) - (2*X %*% V^2) ) ); # shape (n, d)
+ 
+    # gradient w.r.t. w0
```
Comment (j143) [inline scripts/staging/fm.dml]: | grad. of y wrt param  | grad. of  param wrt x   | shape |
| ------------- | ------------- | ------------- |
|  as.matrix(1) |    |(n, 1) |

File: scripts/staging/fm.dml [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    g_w0 = as.matrix(1);  # shape (1, 1)
+ 
+    dw0  = dout * 1 / g_w0 ;  # shape has to be (n, 1)
+    
+    # gradient w.r.t. W
```
Comment (j143-zz) [inline scripts/staging/fm.dml]: | grad. of y wrt param (g_w0) | grad. of  param wrt x   (d_w0) | shape |
| ------------- | ------------- | ------------- |
|  X |  dout * (1 / g_W)  |(n, d) |

File: scripts/staging/fm.dml [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    g_W = X ; # shape (n, d) 
+    
+    dW  = dout * (1 / g_W) ; # shape has to be (d, 1)
+    
+    # gradient w.r.t. V
```
Comment (j143-zz) [inline scripts/staging/fm.dml]: | grad. of y wrt param (g_V) | grad. of  param wrt x   (d_V) | shape |
| ------------- | ------------- | ------------- |
| t(X) %\*% (V %\*% X) - (V %\*% X^2)  |  dout  * 1 / ( g_V ) |(d, k) |

File: scripts/staging/fm.dml [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    *
+    * Outputs:
+    *  - dX : the gradient of loss function w.r.t  X, of shape (n, d).
+    *  - dw0: the gradient of loss function w.r.t w0, of shape (1,).
+    *  - dW : the gradient of loss function w.r.t  W, of shape (d, 1). 
+    *  - dV : the gradient of loss function w.r.t  V, of shape (d, k).
+    */
+    
+    # gradient of y (model) w.r.t. X (input)
+    dX  = ( W + 0.5 * rowSums( (2*X %*% V^2) - (2*X %*% V^2) ) ); # shape (n, d)
```
Comment (j143-zz) [inline scripts/staging/fm.dml]: gradient of model wrt input
```
w0 -> 0 # first term
``` 
```
X %*% W -> W # second term
```
```
rowSums( (X %*% V)^2 - (X^2 %*% V^2) ) -> rowSums( (2*X %*% V^2) - (2*X %*% V^2) ) # third term
```

**Pull request closed without merging**